file_util/src/file.cc

```diff
@@ -0,0 +1,123 @@
+#include <leatherman/file_util/file.hpp>
+#include <boost/nowide/fstream.hpp>
+#include <boost/nowide/cenv.hpp>
+#include <boost/system/error_code.hpp>
+#include <sstream>
+#include <leatherman/logging/logging.hpp>
+
+using namespace std;
+
+namespace leatherman { namespace file_util {
+
+namespace boost_error = boost::system::errc;
+namespace boost_file = boost::filesystem;
+
+    bool each_line(string const& path, function<bool(string&)> callback)
+    {
+        boost::nowide::ifstream in(path.c_str());
+        if (!in) {
+            return false;
+        }
+
+        string line;
+        while (getline(in, line)) {
+            if (!callback(line)) {
+                break;
+            }
+        }
+        return true;
+    }
+
+    string read(string const& path)
+    {
+        string contents;
+        if (!read(path, contents)) {
+            return {};
+        }
+        return contents;
+    }
+
+    bool read(string const& path, string& contents)
+    {
+        boost::nowide::ifstream in(path.c_str(), ios::in | ios::binary);
+        ostringstream buffer;
+        if (!in) {
+            return false;
+        }
+        buffer << in.rdbuf();
+        contents = buffer.str();
+        return true;
+    }
+
+    bool file_readable(const std::string &file_path) {
+        bool exists { false };
+        if (file_path.empty()) {
+            LOG_WARNING("file path is an empty string");
+        } else {
+            boost::nowide::ifstream file_stream(file_path.c_str());
+            exists = file_stream.good();
+            file_stream.close();
+        }
+        return exists;
+    }
+
+    void atomic_write_to_file(const std::string &text,
+                              const std::string &file_path,
+                              std::ios_base::openmode mode) {
+        boost::nowide::ofstream ofs;
+        std::string tmp_name = file_path + "~";
+        ofs.open(tmp_name.c_str(), mode);
+        if (!ofs.is_open()) {
+            throw boost_file::filesystem_error { "failed to open " + file_path,
+                                                        boost_error::make_error_code(
+                                                                boost_error::io_error) };
+        }
+        ofs << text;
+        ofs.close();
+        rename(tmp_name.data(), file_path.data());
+    }
+
+    std::string tilde_expand(std::string path) {
```

Comment:
Every test of this method is failing on Windows due to a "basic_string::_S_construct null not valid" exception. Is there any way to get more information on where this exception is coming from when running the tests?